### ISSUE TITLE
Support for key/pair authentication?

### ISSUE THREAD

**Issue by c-tanner** [appsfortableau/laravel-pdo-odbc#38]
Snowflake will begin requiring MFA on all `USER` type accounts by Nov 2025. `SERVICE` type accounts do not require MFA, but require key/pair authentication.

Are there plans to support this?

https://www.snowflake.com/en/blog/blocking-single-factor-password-authentification/

**Comment by yoramdelangen** [comment]
Yes, we should have support for it! Are your using the ODBC setup of Snowflake or Snowflake PDO?

**Comment by c-tanner** [comment]
Awesome! We're using Snowflake PDO

**Comment by c-tanner** [comment]
Adding a recent email from Snowflake with more granular timelines on this rollout, and mitigation strategies in case they help other folks:

> As part of our continued effort to improve our customer security, we recently [announced](https://go.snowflake.net/MjUyLVJGTy0yMjcAAAGYJH94AJ-9T8zpFKCC4VsksOO15IvT8cavd9WIo7MPTr3kS_HKR7CTAlVj6pACSfzrA4v10YA=) enforcement of MFA by default for new Snowflake accounts. This change will be introduced as part of BCR Bundle 2024_08 and your account will be impacted by this change. This bundle is expected to be deployed for testing during the week of October 7th, 2024 in release 8.38.  Please note dates are subject to change.

Timeline:

> **Opt-in Period (Oct 2024-Early Jan 2025)**: During this period, in order for new Snowflake accounts to have the new built-in authentication policy, you must enable the BCR bundle. This policy will not apply to any new Snowflake account created before you enable the bundle.  
**Enabled by default (Early Jan 2025 to Early Feb 2025)**: All new Snowflake accounts will have the new built-in authentication policy by default unless your Snowflake Account Admin explicitly opts out of BCR 2024_08.
**GA - built-in policy enforced (Early Feb 2025)**: All new Snowflake accounts will have the new built-in authentication policy. Your Snowflake Account Admin will not be able to opt-out. Your Snowflake Account Admin can disable the new built-in authentication by [creating a custom authentication policy](https://go.snowflake.net/MjUyLVJGTy0yMjcAAAGYJH94ACa2UVs2Msq1aRK2BYoyXxA6-ACh-mUBzKNItXuiJiuj71-FV1CEQoHJPmJQuLUpZEs=).
**MFA policy for existing users**: Over the next few quarters, Snowflake is looking to extend this new built-in authentication policy to all existing accounts.

The current workaround is to `ALTER USER [user] SET TYPE = LEGACY_SERVICE`. This is discouraged, but hopefully only necessary temporarily until we have key/pair authentication here.

**Comment by lbattaglioli2000** [comment]
Just curious, have any advancements been made (or plan to be made) here?
